File: blocosAlternados.py
##
# AUTOR: GUSTAVO SIEBRA
# DATA: 11/05/2020
# LOCAL: Fortaleza
# 
# METODO DOS BLOCOS ALTERNADOS

#import math
import numpy as np
import matplotlib.pyplot as plt

def blocosAlternados(dt, Tr, duracao):

    l = duracao/dt
    n = int(l)
    
    index =  [0 for x in range(1) for y in range(int(n))] 
    td =  [0 for x in range(1) for y in range(int(n))] 
    intensidade =  [[0 for x in range(1)] for y in range(int(n))] 
    p =  [0 for x in range(1) for y in range(int(n))] 
    altura =  [0 for x in range(1) for y in range(int(n))] 
    hietograma =  [0 for x in range(1) for y in range(int(n))] 
    posicao =  [0 for x in range(1) for y in range(int(n))] 
    hieto =  [0 for x in range(1) for y in range(int(n))] 
     
    result = [[0 for x in range(8)] for y in range(int(n))] 
    
    for j in range(int(n)):
        if j == 0:
            index[j] = 0
            td[j] = dt
        else:
            index[j] = index[j-1] + 1
            td[j] = td[j-1] + dt
        
        # equacao da chuva de Fortaleza
        #intensidade[j] = (2345.29*(math.pow(Tr, 0.173)))/(math.pow((td[j]+28.31),0.904))
        intensidade[j] = (2345.29*(Tr** 0.173))/((td[j]+28.31)**0.904)
        
        # calculo da precipitacao
        p[j]= intensidade[j] * (td[j]/60)
    
        if j == 0:
            altura[j] = p[j];
        else:
            altura[j] = p[j] - p[j-1]   
        
        # hietograma nao balanceado
        hietograma[j] = (altura[j]/dt) * 60 # mm/h
        
    ## BALANCEAMENTO DO HIETOGRAMA (IMPORTANTE)
    # verifica se e impar ou par
    if n%2 == 0:
        #sprintf('par');
        for i in range(0,int(n/2)):
            posicao[i] = n-((2*i)+2)
    
        for i in range((int(n/2)),int(n)):
            posicao[i] = (2*i)-n+1
    
    else:
        #sprintf('é impar')
        for i in range(0,int(n/2)):
            posicao[i] = n-((2*i)+2)
    
        for i in range((int(n/2)+1),int(n)):
            posicao[i] = (2*i)-n+1
        
    for m in range(0,int(n)):
        pos = posicao[m]
        hieto[m] = hietograma[pos]
    
        result[m][0] = index[m]
        result[m][1] = td[m]
        result[m][2] = round(intensidade[m],2)
        result[m][3] = round(p[m],2)
        result[m][4] = round(altura[m],2)
        result[m][5] = round(hietograma[m],2)
        result[m][6] = posicao[m]
        result[m][7] = round(hieto[m],2)
        
    #print(result)
    
    ## salvar em arquivo csv
    np.savetxt("resultado.csv", result, delimiter=",", fmt='%10.2f') 
    
    ## abrindo arquivo salvo
    a = open("resultado.csv", 'r')# open file in read mode  
    print(a.read()) 
    
    ## plotar os resultados
    plt.bar(td, hietograma)
    plt.xlabel('tempo (min)')
    plt.ylabel('precipitação (mm)')
    plt.title('hietograma')
    plt.show()
    plt.bar(td, hieto)
    plt.xlabel('tempo (min)')
    plt.ylabel('precipitação (mm)')
    plt.title('hietograma Balanceado')

## chamada da funcao
bloco = blocosAlternados(6, 50, 60)

File: blocosAlternados.m
%%
% AUTOR: GUSTAVO SIEBRA
% DATA: 24/04/2020
% LOCAL: MACAPA
% 
% METODO DOS BLOCOS ALTERNADOS
%% LIMPANDO VARIAVEIS
%
clc
clear all
close all
%% VARIAVEIS
%
dt = 6; %intervalo de discretizacao, em minutos.
Tr = 50; %Periodo de Retorno, em anos.
duracao = 60; %duracao do evento
%% INICIALIZACAO
%
n = duracao/dt; %numero de elementos do vetor chuva

indice(n,1) = 0;
td(n,1) = 0; %tempo de duracao, minutos
intensidade(n,1) = 0; %recebe o resultado da intensidade, mm/h
p(n,1) = 0; %precipitacao em milimetros
altura(n,1) = 0; %diferenca de laminacao da chuva em mm
hietograma(n,1) = 0; %hieograma bruto
posicao(n,1) = 0; %posicao ordenada
hieto_ba(n,1) = 0; %hietograma ordenado
resultado(n,7) = 0; %variavel que recebe todos os valores
%% NUCLEO
%
for m=1:n
    
    if m==1
        indice(m,1) = 1;
        td(m,1) = dt;
    else 
        indice(m,1) = indice(m-1,1) + 1;
        td(m,1) = dt + td(m-1);
    end
    
    %if m==1
    %    td(m,1) = dt;
    %else 
    %    td(m,1) = dt + td(m-1);
    %end
    
    % equacao da chuva de Fortaleza
    intensidade(m,1) = (2345.29*(Tr^0.173))/((td(m)+28.31)^0.904); 
    
    %calculo da precipitacao
    p(m,1) = intensidade(m) * (td(m)/60);
    
    %calculo da altura da lamina
    if m == 1
        altura(m) = p(m);
    else
        altura(m) = p(m) - p(m-1);
    end
    
    %hietograma nao balanceado
    hietograma(m,1) = (altura(m)/dt) * 60; % mm/h
end
%% BALANCEAMENTO DO HIETOGRAMA (IMPORTANTE)
%
%maximo = max(hietograma);

% verifica se e impar ou par
if mod(n,2) == 0
    sprintf('par');
    for i=1:n/2
        posicao(i,1) = n-((2*i)-1);
        %posicao(i,1) = n-((2*i)-2); %outra forma de calcular
    end

    for i=(n/2)+1:n
        posicao(i,1) = (2*i)-n;
        %posicao(i,1) = ((2*i)-1)-n; %outra forma de calcular
    end
else
    sprintf('é impar');
    for i=1:n/2
        posicao(i,1) = n-((2*i)-1);
    end

    for i=(n+1)/2:n
        posicao(i,1) = (2*i)-n;
    end
end
%% HIETOGRAMA BALANCEADO
%
for m=1:n
    hieto_ba = hietograma(posicao);
end
%% UNINDO TODOS OS RESULTADOS
%
resultado(:,1) = indice; %tempo de duracao, minutos
resultado(:,2) = td; %tempo de duracao, minutos
resultado(:,3) = intensidade; %recebe o resultado da intensidade, mm/h
resultado(:,4) = p; %precipitacao em milimetros
resultado(:,5) = altura; %diferenca de laminacao da chuva em mm
resultado(:,6) = hietograma; %hieograma bruto
resultado(:,7) = posicao; %posicao ordenada
resultado(:,8) = hieto_ba; %hietograma ordenado

resultado %escreve na tela

bar(td,hietograma) %mostra o grafico do tempo x hietograma
ylabel('precipitação (mm)');
xlabel('tempo (min)');
title('Hietograma')
figure
bar(td,hieto_ba) %mostra o grafico do tempo x hietograma
ylabel('precipitação (mm)');
xlabel('tempo (min)');
title('Hietograma')